File: java/base/src/main/vsdk/toolkit/common/statistics/RenderingStatistics.java
package vsdk.toolkit.common.statistics;

import vsdk.toolkit.common.VSDK;

/**
 * Centralized primitive and intersection counters used by rendering flows.
 */
public final class RenderingStatistics
{
    private static final int primitiveCount[];
    private static final int intersectionCount[];

    static {
        primitiveCount = new int[VSDK.PRIMITIVE_TYPE_COUNT];
        intersectionCount = new int[VSDK.INTERSECTION_TYPE_COUNT];
        resetPrimitiveCounters();
        resetIntersectionCounters();
    }

    private RenderingStatistics()
    {
    }

    public static void resetPrimitiveCounters()
    {
        int i;
        for ( i = 0; i < VSDK.PRIMITIVE_TYPE_COUNT; i++ ) {
            primitiveCount[i] = 0;
        }
    }

    public static void resetIntersectionCounters()
    {
        int i;
        for ( i = 0; i < VSDK.INTERSECTION_TYPE_COUNT; i++ ) {
            intersectionCount[i] = 0;
        }
    }

    public static void accumulatePrimitiveCount(int type, int count)
    {
        primitiveCount[type] += count;
    }
}

File: typescript/base/src/main/vsdk/toolkit/common/statistics/RenderingStatistics.ts
import { VSDK } from "../VSDK.js";

/** Centralized primitive and intersection counters. */
export class RenderingStatistics {
  private static readonly primitiveCount = new Int32Array(VSDK.PRIMITIVE_TYPE_COUNT);
  private static readonly intersectionCount = new Int32Array(VSDK.INTERSECTION_TYPE_COUNT);
  public static resetPrimitiveCounters(): void { this.primitiveCount.fill(0); }
  public static resetIntersectionCounters(): void { this.intersectionCount.fill(0); }
  public static accumulatePrimitiveCount(type: number, count: number): void { if (!Number.isInteger(type) || type < 0 || type >= this.primitiveCount.length) throw new RangeError(`Primitive type out of bounds: ${type}`); this.primitiveCount[type]! += count; }
}
